service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/AllergyStructureExtractor.java

```diff
@@ -0,0 +1,89 @@
+package uk.nhs.adaptors.gp2gp.ehr.mapper;
+
+import static uk.nhs.adaptors.gp2gp.ehr.utils.CodeableConceptMappingUtils.extractTextOrCoding;
+import static uk.nhs.adaptors.gp2gp.ehr.utils.DateFormatUtil.toHl7Format;
+
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.Collectors;
+
+import org.apache.commons.lang3.StringUtils;
+import org.hl7.fhir.dstu3.model.AllergyIntolerance;
+import org.hl7.fhir.dstu3.model.DateTimeType;
+import org.hl7.fhir.dstu3.model.Extension;
+import org.hl7.fhir.dstu3.model.StringType;
+
+import uk.nhs.adaptors.gp2gp.ehr.utils.CodeableConceptMappingUtils;
+import uk.nhs.adaptors.gp2gp.ehr.utils.DateFormatUtil;
+
+public class AllergyStructureExtractor {
+    private static final String ALLERGY_REASON_ENDED = "Reason Ended: ";
+    private static final String ALLERGY_REASON_END_URL = "reasonEnded";
+    private static final String ALLERGY_END_DATE_URL = "endDate";
+    private static final String ALLERGY_NO_INFO = "No information available";
+    private static final String REACTION = "Reaction %S";
+    private static final String REACTION_DESCRIPTION = " Description: ";
+    private static final String REACTION_SEVERITY = " Severity: ";
+    private static final String REACTION_EXPOSURE_ROUTE = " Exposure Route: ";
+    private static final String REACTION_MANIFESTATIONS = " Manifestation(s): ";
+    private static final String COMMA = ", ";
+
+    public static String extractReasonEnd(Extension extension) {
+        var reasonEnd = extension.getExtension()
```

Comment:
the if statement below could be done by additional filter + map :

```
     return extension.getExtension().stream()
            .filter(value -> value.getUrl().equals(ALLERGY_REASON_END_URL))
            .findFirst()
            .map(Extension::getValue)
            .map(reason -> ((StringType) reason).getValueAsString())
            .filter(reasonEnd -> !ALLERGY_NO_INFO.equals(reasonEnd))
            .map(reasonEnd -> ALLERGY_REASON_ENDED + reasonEnd)
            .orElse(StringUtils.EMPTY);
```

should to the same.

Reply:
do we only want to find the first?

Reply:
Allergy End has cardinality of 0..1, but Extension in HAPI is an array thus we only want to find the first.